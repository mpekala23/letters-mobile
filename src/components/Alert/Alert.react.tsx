import React, { createRef } from 'react';
import {
  Text,
  View,
  ViewStyle,
  TextStyle,
  TouchableOpacity,
  Animated,
  Platform,
} from 'react-native';
import { Colors, Typography } from '@styles';
import Button from '../Button/Button.react';
import Styles, { GRAY_BACK } from './Alert.styles';
import { setStatusBackground } from '../Statusbar/Statusbar.react';

interface AlertButton {
  text?: string;
  reverse?: boolean;
  onPress?: () => void;
  textStyle?: TextStyle | TextStyle[];
  containerStyle?: ViewStyle | ViewStyle[];
}

interface AlertInfo {
  title?: string;
  message?: string;
  buttons?: AlertButton[];
}

interface State {
  current: AlertInfo | null;
  opacity: Animated.Value;
}

class Alert extends React.Component<Record<string, unknown>, State> {
  constructor(props: Record<string, unknown>) {
    super(props);
    this.state = {
      current: null,
      opacity: new Animated.Value(0.0),
    };
    this.setCurrent = this.setCurrent.bind(this);
  }

  setCurrent(info: AlertInfo | null): void {
    if (info) this.setState({ current: info });
    Animated.timing(this.state.opacity, {
      toValue: info ? 1.0 : 0.0,
      duration: info ? 0 : 80,
      useNativeDriver: true,
    }).start(() => {
      this.setState({ current: info });
    });
  }

  render(): JSX.Element {
    if (!this.state.current) return <View />;
    return (
      <Animated.View
        style={[Styles.trueBackground, { opacity: this.state.opacity }]}
      >
        <TouchableOpacity
          style={Styles.trueBackground}
          onPress={() => {
            this.setCurrent(null);
            setStatusBackground('white');
          }}
          activeOpacity={1.0}
        >
          <View style={Styles.alertBackground}>
            <Text
              style={[
                Typography.FONT_BOLD,
                { fontSize: 20, textAlign: 'center', marginBottom: 18 },
              ]}
            >
              {this.state.current.title}
            </Text>
            {this.state.current.message && (
              <>
                <Text
                  style={[
                    Typography.FONT_REGULAR,
                    {
                      fontSize: 16,
                      textAlign: 'center',
                      color: Colors.GRAY_DARK,
                    },
                  ]}
                >
                  {this.state.current.message}
                </Text>
                <View style={{ height: 32 }} />
              </>
            )}
            {this.state.current.buttons &&
              this.state.current.buttons.map((button) => {
                return (
                  <Button
                    key={button.text}
                    buttonText={button.text}
                    reverse={button.reverse}
                    onPress={() => {
                      if (button.onPress) button.onPress();
                      this.setCurrent(null);
                      setStatusBackground('white');
                    }}
                    textStyle={button.textStyle}
                    containerStyle={
                      button.containerStyle
                        ? button.containerStyle
                        : { width: '100%' }
                    }
                  />
                );
              })}
          </View>
        </TouchableOpacity>
      </Animated.View>
    );
  }
}

const alertRef = createRef<Alert>();
const AlertInstance = (): JSX.Element => <Alert ref={alertRef} key="Alert" />;

export function popupAlert(pop: AlertInfo): void {
  setStatusBackground(Platform.OS === 'ios' ? GRAY_BACK : 'rgb(145,145,145)');
  if (alertRef.current) {
    alertRef.current.setCurrent({
      title: pop.title,
      message: pop.message,
      buttons: pop.buttons,
    });
  } else {
    setStatusBackground('white');
  }
}

export default AlertInstance;
