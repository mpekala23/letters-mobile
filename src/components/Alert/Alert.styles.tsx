import { StyleSheet } from 'react-native';
import { WINDOW_HEIGHT, WINDOW_WIDTH } from '@utils';

export const GRAY_BACK = 'rgba(0,0,0,0.25)';

export default StyleSheet.create({
  trueBackground: {
    position: 'absolute',
    width: WINDOW_WIDTH,
    height: WINDOW_HEIGHT,
    backgroundColor: GRAY_BACK,
    justifyContent: 'center',
    alignItems: 'center',
    zIndex: 999,
    elevation: 999,
  },
  alertBackground: {
    backgroundColor: 'white',
    width: '85%',
    borderRadius: 6,
    justifyContent: 'center',
    alignItems: 'center',
    padding: 20,
  },
});
