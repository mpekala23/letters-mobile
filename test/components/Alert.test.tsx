import Alert, { popupAlert } from '@components/Alert/Alert.react';
import { render, toJSON, fireEvent } from '@testing-library/react-native';

const setup = () => {
  return {
    ...render(Alert()),
  };
};

describe('Alert component', () => {
  it('should match snapshot', () => {
    const { container } = setup();
    const tree = toJSON(container);
    expect(tree).toMatchSnapshot();
  });

  it('should implement alert titles', () => {
    const { getByText } = setup();
    popupAlert({
      title: 'Title',
      message: '',
      buttons: [],
    });
    expect(getByText('Title')).toBeDefined();
  });

  it('should implement alert messages', () => {
    const { getByText } = setup();
    popupAlert({
      title: '',
      message: 'Message',
      buttons: [],
    });
    expect(getByText('Message')).toBeDefined();
  });

  it('should implement alert buttons', () => {
    const { getByText } = setup();
    const dummy = jest.fn();
    popupAlert({
      title: '',
      message: '',
      buttons: [
        {
          text: 'Button',
          reverse: true,
          onPress: dummy,
        },
      ],
    });
    const button = getByText('Button');
    expect(button).toBeDefined();
    fireEvent.press(button);
    expect(dummy).toHaveBeenCalledTimes(1);
  });
});
