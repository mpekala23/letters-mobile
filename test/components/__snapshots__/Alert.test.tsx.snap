// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert component should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View />
</View>
`;
